### PR TITLE
[Processing] Fix wrong result of Rescale Raster when input data type is not Float32 (fixes #41084)

### LINKED ISSUE #41084: [Processing] Rescale Raster gives bad results.
The Rescale Raster tool gives nonsensical results in some instances.

I have a raster (float64) whose value are ranging from -1 to 255.
image

As those should represent a direction (originally from r.terraflow ) I want to turn them into quadrants.

When I process the data to rescale the values between 0 and 7 or 1 and 8.

In the resulting layer I seem to get an overflow and values are either between e-20 and 0 while some others are up in 7e+37
image

Tested on 3.16.1 and 3.16.3 ( the alg is inchanged so any change would be surprisig).

Works fine on raster with a positive range.

### PR DESCRIPTION
**Pull request by rouault**

### CODE REVIEW

**Review by alexbruy — APPROVED** [review]

**Comment by roya0045** [comment]
Thank you for looking into and fixing this Even!

File: src/analysis/processing/qgsalgorithmrescaleraster.cpp [hunk 1]
```diff
@@ -143,9 +153,10 @@ QVariantMap QgsRescaleRasterAlgorithm::processAlgorithm( const QVariantMap &para
   int iterTop = 0;
   int iterCols = 0;
   int iterRows = 0;
-  std::unique_ptr< QgsRasterBlock > block;
-  while ( iter.readNextRasterPart( mBand, iterCols, iterRows, block, iterLeft, iterTop ) )
+  std::unique_ptr< QgsRasterBlock > inputBlock;
+  while ( iter.readNextRasterPart( mBand, iterCols, iterRows, inputBlock, iterLeft, iterTop ) )
   {
+    std::unique_ptr< QgsRasterBlock > outputBlock( new QgsRasterBlock( destProvider->dataType( 1 ), iterCols, iterRows ) );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmrescaleraster.cpp:159]: Can you explain this change? It's causing the block to be created twice...

Comment (rouault) [reply src/analysis/processing/qgsalgorithmrescaleraster.cpp:159]: we need separate QgsRasterBlock for input and ouput as they might be of different data types. The issue here was that when we had a Float64 dataset, block->setValue() was written the value as a double, but the code called by destProvider->writeBlock( block.get(), mBand, iterLeft, iterTop ); assumed the block was of the data type of destProvider, that is Float32, hence chaos

Comment (nyalldawson) [reply src/analysis/processing/qgsalgorithmrescaleraster.cpp:159]: Good point, thanks!

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**